[DCP] Support partial load

LucasLLC opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122829

Adds ability to load a subset of keys directly from a checkpoint, avoiding the need to initialize state dict first

Differential Revision: [D55441391](https://our.internmc.facebook.com/intern/diff/D55441391/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

LucasLLC committed:
[DCP] Support partial load

Adds ability to load a subset of keys directly from a checkpoint, avoiding the need to initialize state dict first

Differential Revision: [D55441391](https://our.internmc.facebook.com/intern/diff/D55441391/)

[ghstack-poisoned]

LucasLLC commented on torch/distributed/checkpoint/default_planner.py:
```diff
@@ -216,6 +216,47 @@ def transform_tensor(self, read_item: ReadItem, tensor: torch.Tensor):
         return narrow_tensor_by_index(tensor, read_item.dest_offsets, read_item.lengths)
 
 
+class _EmptyStateDictLoadPlanner(DefaultLoadPlanner):
```
I had to move this out of format utils to avoid a circular dependency. I'm very open to suggestions on where a better home might be for this class

wz337 commented:
Looks like `test/distributed/checkpoint/test_partial_load.py` is empty?

fegin commented:
`_load_state_dict_from_keys` seems to violate the fundamental philosophy of DCP. 1. It does not support resharding, 2. It does not support in-place loading. Why do we actually need such an API?

fegin reviewed (changes requested):
_load_state_dict_from_keys seems to violate the fundamental philosophy of DCP. 1. It does not support resharding, 2. It does not support in-place loading. Why do we need such an API?

LucasLLC commented:
> _load_state_dict_from_keys seems to violate the fundamental philosophy of DCP. 1. It does not support resharding, 2. It does not support in-place loading. Why do we need such an API?

This was answered offline:
The use case is that researchers often want to just load specific modules and inspect them on the fly. E.g. someone shared a checkpoint file with you, and we would like to load it into a notebook and inspect it on the fly. I think we can consider this a “utility” for that use case.

LucasLLC committed:
Update on "[DCP] Support partial load"


Adds ability to load a subset of keys directly from a checkpoint, avoiding the need to initialize state dict first

Differential Revision: [D55441391](https://our.internmc.facebook.com/intern/diff/D55441391/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

fegin commented on torch/distributed/checkpoint/state_dict_loader.py:
```diff
@@ -215,3 +216,66 @@ def read_data():
+
+
+def _load_state_dict_from_keys(
+    keys: Optional[set],
```
Why is this optional? If users pass None, is this API going to load everything? If so, we may want to mention that in the docstring. If this API will do nothing, then I suggest that we should simply not support the None case.

fegin commented on torch/distributed/checkpoint/state_dict_loader.py:
```diff
@@ -215,3 +216,66 @@ def read_data():
+    process_group: Optional[dist.ProcessGroup] = None,
+) -> Dict[str, Any]:
+    """
+    Load only the specified keys from the checkpoint.
```
We should also mention that the checkpoint will not be sharded and all the ranks, if distributed is initialized, are going to load the full checkpoint.

LucasLLC committed:
Update on "[DCP] Support partial load"


Adds ability to load a subset of keys directly from a checkpoint, avoiding the need to initialize state dict first

Differential Revision: [D55441391](https://our.internmc.facebook.com/intern/diff/D55441391/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

LucasLLC commented:
@pytorchbot merge
